plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -127,3 +140,36 @@ counted out already so we should not immediately encounter those in `VBody`.
 Also, we currently reject `Constant` (has acceptable cost but not acceptable size).
 We may want to check their sizes instead of just rejecting them.
 -}
+
+-- | A list of `LamAbs` and `TyAbs`, in order, of a let-binding.
+type TermOrTypeOrder = [TermOrType]
+
+-- | Datatype capturing both terms and types.
+data TermOrType =
+    MkTerm | MkType
+    deriving stock (Eq, Show)
+
+instance Pretty TermOrType where
+  pretty = viaShow
+
+-- | Counts the type and term lambdas in the RHS of a binding and returns an ordered list
+countLam ::
+    Term tyname name uni fun a -- ^ the RHS of the let binding
+    -> TermOrTypeOrder
+countLam = countLamInner []
```

Comment:
This is a left fold, which is inefficient: you need to append to the list (`lamStack <> [MkTerm])`, and the first element of the list won't be available until the whole function is done.

A right fold is both shorter and more efficient:

```haskell
countLam = \case
  LamAbs _ _ _ body -> MkTerm : countLam body
  TyAbs _ _ _ body -> MkType : countLam body
  _ -> countLam body